core/src/main/java/com/microsoft/applicationinsights/internal/config/connection/EndpointProvider.java

```diff
@@ -0,0 +1,95 @@
+package com.microsoft.applicationinsights.internal.config.connection;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.microsoft.applicationinsights.internal.config.connection.ConnectionString.Defaults;
+import org.apache.http.client.utils.URIBuilder;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class EndpointProvider {
+    @VisibleForTesting static final String INGESTION_URI_PATH = "v2/track";
+    @VisibleForTesting static final String LIVE_URI_PATH = "QuickPulseService.svc";
+    @VisibleForTesting static final String API_PROFILES_APP_ID_URI_PREFIX = "api/profiles/"; // <base uri, with host>/api/profiles/<ikey>/appid
+    @VisibleForTesting static final String API_PROFILES_APP_ID_URI_SUFFIX = "/appId";
+
+    private final AtomicReference<URI> ingestionEndpoint = new AtomicReference<>();
+    private final AtomicReference<URI> ingestionEndpointURL = new AtomicReference<>();
+    private final AtomicReference<URI> liveEndpointURL = new AtomicReference<>();
+    private final AtomicReference<URI> profilerEndpoint = new AtomicReference<>();
+    private final AtomicReference<URI> snapshotEndpoint = new AtomicReference<>(); // TODO is this one needed?
```

Comment:
i think use `volatile` instead of `AtomicReference`, since not doing any atomic operations on them

Reply:
In this case, I think it's equivalent since URI is immutable.